File: experiments/java-sdk/src/main/java/com/datadog/llmobs/experiments/ExperimentsClient.java
package com.datadog.llmobs.experiments;

import com.datadog.api.client.ApiClient;
import com.datadog.api.client.v2.api.LlmObservabilityApi;
import com.datadog.api.client.v2.api.LlmObservabilityApi.ListLLMObsDatasetsOptionalParameters;
import com.datadog.api.client.v2.model.LLMObsDatasetDataResponse;
import com.datadog.api.client.v2.model.LLMObsDatasetsResponse;
import com.datadog.llmobs.experiments.internal.DirectPost;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.datadog.api.client.v2.model.LLMObsProjectDataAttributesRequest;
import com.datadog.api.client.v2.model.LLMObsProjectDataRequest;
import com.datadog.api.client.v2.model.LLMObsProjectRequest;
import com.datadog.api.client.v2.model.LLMObsProjectType;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.introspect.AnnotatedMember;
import com.fasterxml.jackson.databind.introspect.JacksonAnnotationIntrospector;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Entry point for the Datadog LLM Experiments Java SDK.
 *
 * <p>Wraps {@code datadog-api-client-java}'s {@link ApiClient} and {@link LlmObservabilityApi},
 * applying the LLM Observability auth, site, and unstable-operation opt-ins the SDK needs.
 *
 * <p>Construct via {@link #builder()}.
 */
public final class ExperimentsClient {

    private static final String[] UNSTABLE_OPERATIONS = new String[] {
        "v2.createLLMObsProject",
        "v2.listLLMObsProjects",
        "v2.createLLMObsDataset",
        "v2.listLLMObsDatasets",
        "v2.createLLMObsDatasetRecords",
        "v2.updateLLMObsDatasetRecords",
        "v2.listLLMObsDatasetRecords",
        "v2.deleteLLMObsDatasetRecords",
        "v2.createLLMObsExperiment",
        "v2.createLLMObsExperimentEvents",
    };

    private static final ObjectMapper JSON = new ObjectMapper();

    private final ApiClient apiClient;
    private final LlmObservabilityApi api;
    private final String projectName;
    private final String site;
    private final String apiKey;
    private final String appKey;
    private volatile String cachedProjectId;

    private ExperimentsClient(Builder b) {
        // Use a fresh ApiClient rather than getDefaultApiClient(): the latter eagerly reads
        // DD_SITE from the environment and validates it against a fixed allow-list of public
        // sites, which rejects staging hosts like datad0g.com before we get a chance to
        // configure anything.
        ApiClient ac = new ApiClient();

        // Server index 2 is the "any Datadog deployment" variant with an unrestricted site
        // variable (the public-sites enum lives at index 0). Using index 2 lets us point at
        // staging, custom, or any future site without bumping the client library.
        ac.setServerIndex(2);

        // The LLM Obs endpoints are still being shaped — some response fields the OpenAPI
        // spec marks "required" are actually omitted by the live API (e.g. project.description).
        // Strict deserialization fails those responses. Override Jackson's annotation
        // introspector to claim no property is required, which relaxes response parsing
        // without affecting request serialization. We also disable FAIL_ON_MISSING_CREATOR_PROPERTIES
        // as a belt-and-suspenders. Remove once the spec aligns with the live API.
        ac.getJSON().getMapper()
            .setAnnotationIntrospector(new JacksonAnnotationIntrospector() {
                @Override
                public Boolean hasRequiredMarker(AnnotatedMember m) {
                    return Boolean.FALSE;
                }
            });
        ac.getJSON().getMapper()
            .configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);

        Map<String, String> authKeys = new HashMap<>();
        authKeys.put("apiKeyAuth", b.apiKey);
        authKeys.put("appKeyAuth", b.appKey);
        ac.configureApiKeys(authKeys);

        Map<String, String> serverVars = new HashMap<>();
        serverVars.put("site", b.site);
        ac.setServerVariables(serverVars);

        for (String op : UNSTABLE_OPERATIONS) {
            ac.setUnstableOperationEnabled(op, true);
        }

        this.apiClient = ac;
        this.api = new LlmObservabilityApi(ac);
        this.projectName = b.projectName;
        this.site = b.site;
        this.apiKey = b.apiKey;
        this.appKey = b.appKey;
    }

    // Package-private accessors used by EventsPoster (hand-rolled HTTP workaround).
    // Not public — credentials should not leak beyond the package.
    String apiKey() {
        return apiKey;
    }

    String applicationKey() {
        return appKey;
    }

    /** Generated API surface for any endpoint this SDK does not wrap. */
    public LlmObservabilityApi api() {
        return api;
    }

    /** Create a new {@link Dataset} bound to this client. Not pushed to Datadog until used. */
    public Dataset createDataset(String name) {
        return new Dataset(this, name, null);
    }

    public Dataset createDataset(String name, String description) {
        return new Dataset(this, name, description);
    }

    /**
     * Pull an existing dataset by name. Returns a {@link Dataset} pre-populated with every
     * record (and each record's stable ID) from Datadog. Useful for inspecting datasets the
     * UI / Python SDK created, or for resuming work on a dataset across sessions.
     *
     * @throws IllegalArgumentException if no dataset with that name exists in the project.
     */
    public Dataset pullDataset(String name) {
        Objects.requireNonNull(name, "name");
        String projectId = ensureProjectId();

        // The list endpoint is index-backed and lags briefly behind dataset creation. Retry
        // up to a few times with backoff so a "push immediately followed by pull" pattern
        // (common in examples and tests) works without the caller having to sleep.
        List<LLMObsDatasetDataResponse> hits = null;
        long[] backoffMs = {0L, 1000L, 2000L, 3000L};
        for (long delay : backoffMs) {
            if (delay > 0) {
                try {
                    Thread.sleep(delay);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
            LLMObsDatasetsResponse listResp;
            try {
                listResp = api.listLLMObsDatasets(projectId, new ListLLMObsDatasetsOptionalParameters().filterName(name));
            } catch (Exception e) {
                throw new RuntimeException("Failed to list datasets in project '" + projectName + "': " + e.getMessage(), e);
            }
            hits = listResp.getData();
            if (hits != null && !hits.isEmpty()) {
                break;
            }
        }

        if (hits == null || hits.isEmpty()) {
            throw new IllegalArgumentException("Dataset '" + name + "' not found in project '" + projectName + "' (after retries)");
        }

        LLMObsDatasetDataResponse dsData = hits.get(0);
        String datasetId = dsData.getId();
        String description = dsData.getAttributes() != null && dsData.getAttributes().getDescription() != null
            ? dsData.getAttributes().getDescription()
            : "";

        // Hand-rolled GET for records: the generated LLMObsDatasetRecordDataResponse expects
        // flat fields (getInput, getExpectedOutput, getMetadata directly on the item) but the
        // live API wraps them inside data[].attributes — yet another spec/API mismatch on the
        // read side. Parse JSON manually until the spec catches up.
        //
        // Records have their own eventual-consistency window separate from the dataset list,
        // so retry independently if we see 0 records (could be true empty, or could be lag).
        List<DatasetRecord> records = new ArrayList<>();
        List<String> recordIds = new ArrayList<>();
        for (long delay : backoffMs) {
            if (delay > 0) {
                try {
                    Thread.sleep(delay);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
            String recordsBody = DirectPost.get(
                site,
                "/api/v2/llm-obs/v1/" + projectId + "/datasets/" + datasetId + "/records",
                apiKey,
                appKey
            );
            records.clear();
            recordIds.clear();
            try {
                JsonNode root = JSON.readTree(recordsBody);
                JsonNode dataArray = root.path("data");
                if (dataArray.isArray()) {
                    for (JsonNode item : dataArray) {
                        recordIds.add(item.path("id").asText(""));
                        JsonNode attrs = item.path("attributes");
                        Object input = JSON.treeToValue(attrs.path("input"), Object.class);
                        Object expected = attrs.has("expected_output") && !attrs.path("expected_output").isNull()
                            ? JSON.treeToValue(attrs.path("expected_output"), Object.class)
                            : null;
                        @SuppressWarnings("unchecked")
                        Map<String, Object> metadata = attrs.has("metadata") && !attrs.path("metadata").isNull()
                            ? JSON.treeToValue(attrs.path("metadata"), Map.class)
                            : Collections.emptyMap();
                        records.add(new DatasetRecord(
                            input != null ? input : Collections.emptyMap(),
                            expected,
                            metadata
                        ));
                    }
                }
            } catch (Exception e) {
                throw new RuntimeException("Failed to parse records for dataset '" + name + "': " + e.getMessage(), e);
            }
            if (!records.isEmpty()) {
                break;
            }
        }

        return Dataset.fromExisting(this, name, description, datasetId, projectId, records, recordIds);
    }

    /**
     * Idempotently resolve the project ID for {@link #projectName}. Cached after first call —
     * the backend dedupes projects by name, so repeated calls return the same ID, but caching
     * avoids unnecessary round-trips when the same client is reused across many push/pull/run
     * operations.
     *
     * <p>Package-private: used by {@link Dataset#push()} and {@link Experiment#run()}.
     */
    String ensureProjectId() {
        if (cachedProjectId != null) {
            return cachedProjectId;
        }
        LLMObsProjectRequest req = new LLMObsProjectRequest(
            new LLMObsProjectDataRequest()
                .type(LLMObsProjectType.PROJECTS)
                .attributes(new LLMObsProjectDataAttributesRequest(projectName))
        );
        try {
            this.cachedProjectId = api.createLLMObsProject(req).getData().getId();
        } catch (Exception e) {
            throw new RuntimeException("Failed to create or get project '" + projectName + "': " + e.getMessage(), e);
        }
        return cachedProjectId;
    }

    public String projectName() {
        return projectName;
    }

    public String site() {
        return site;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private String apiKey;
        private String appKey;
        private String site = "datadoghq.com";
        private String projectName;

        public Builder apiKey(String apiKey) {
            this.apiKey = apiKey;
            return this;
        }

        public Builder applicationKey(String appKey) {
            this.appKey = appKey;
            return this;
        }

        public Builder site(String site) {
            this.site = site;
            return this;
        }

        public Builder projectName(String projectName) {
            this.projectName = projectName;
            return this;
        }

        public ExperimentsClient build() {
            Objects.requireNonNull(apiKey, "apiKey is required");
            Objects.requireNonNull(appKey, "applicationKey is required");
            Objects.requireNonNull(projectName, "projectName is required");
            Objects.requireNonNull(site, "site is required");
            return new ExperimentsClient(this);
        }
    }
}

File: experiments/node-sdk/src/ExperimentsClient.ts
import type { v2 } from "@datadog/datadog-api-client";
import { Dataset } from "./Dataset.js";
import { DatasetRecord } from "./DatasetRecord.js";
import { retryWithBackoff } from "./internal/backoff.js";
import { buildLlmObsApi, type HttpLibrary } from "./internal/generatedClient.js";
import type { ApiCredentials } from "./internal/http.js";

/** Options for {@link ExperimentsClient.pullDataset}. */
export interface PullDatasetOptions {
  /**
   * Wait (with exponential backoff) until at least this many records are
   * readable from the backend. Use when pulling right after a push, to absorb
   * read-after-write lag. If omitted, the pull only waits for the dataset itself
   * to become visible.
   */
  expectedRecordCount?: number;
  /** Maximum total time to wait for the dataset/records to appear. Default 30s. */
  maxWaitMs?: number;
}

export interface ExperimentsClientOptions {
  apiKey: string;
  applicationKey: string;
  /** Bare site host. Defaults to "datadoghq.com". */
  site?: string;
  projectName: string;
  /**
   * @internal — inject a custom `HttpLibrary` into the generated client. Used by
   * the test suite to mock transport; not part of the public contract.
   */
  httpApi?: HttpLibrary;
}

/**
 * Auth + site + project; the factory for everything else.
 *
 * Object-argument constructor, mirroring the Java `ExperimentsClient.builder()`
 * surface. A static `builder()` is also provided so code reads the same as the
 * Java/other-language examples.
 */
export class ExperimentsClient {
  private readonly creds: ApiCredentials;
  private readonly _projectName: string;
  private readonly _api: v2.LLMObservabilityApi;
  private cachedProjectId: string | null = null;

  constructor(options: ExperimentsClientOptions) {
    if (!options.apiKey) throw new Error("apiKey is required");
    if (!options.applicationKey) throw new Error("applicationKey is required");
    if (!options.projectName) throw new Error("projectName is required");
    this.creds = {
      apiKey: options.apiKey,
      appKey: options.applicationKey,
      site: options.site ?? "datadoghq.com",
    };
    this._projectName = options.projectName;
    this._api = buildLlmObsApi(
      this.creds.apiKey,
      this.creds.appKey,
      this.creds.site,
      options.httpApi,
    );
  }

  /** Fluent builder, parity with the Java `ExperimentsClient.builder()`. */
  static builder(): ExperimentsClientBuilder {
    return new ExperimentsClientBuilder();
  }

  projectName(): string {
    return this._projectName;
  }

  site(): string {
    return this.creds.site;
  }

  /** @internal — used by Dataset/Experiment for the hand-rolled HTTP calls (W1/W2/status). */
  apiKey(): string {
    return this.creds.apiKey;
  }

  /** @internal */
  applicationKey(): string {
    return this.creds.appKey;
  }

  /** @internal — credentials bundle for the internal HTTP helpers. */
  credentials(): ApiCredentials {
    return this.creds;
  }

  /**
   * Escape hatch: the underlying generated `LLMObservabilityApi`.
   *
   * Parity with the Java SDK's `api()`. Use it to reach LLM Obs endpoints the six
   * public types do not model (annotation queues, custom evals, deletes, etc.).
   */
  api(): v2.LLMObservabilityApi {
    return this._api;
  }

  /** Create a local dataset buffer. Pushed remotely on first `experiment.run()`. */
  createDataset(name: string, description = ""): Dataset {
    return new Dataset(this, name, description);
  }

  /**
   * Resolve the project id for `projectName`, creating the project if it does not
   * already exist. The create endpoint is get-or-create on name, so a repeated
   * call simply returns the existing id. Cached after the first resolution.
   */
  async ensureProjectId(): Promise<string> {
    if (this.cachedProjectId !== null) return this.cachedProjectId;
    try {
      const resp = await this._api.createLLMObsProject({
        body: { data: { type: "projects", attributes: { name: this._projectName } } },
      });
      this.cachedProjectId = String(resp?.data?.id ?? "");
      return this.cachedProjectId;
    } catch (err) {
      throw new Error(
        `Failed to create or get project '${this._projectName}': ${(err as Error).message}`,
      );
    }
  }

  /**
   * Pull an existing dataset by name (with its records) from the current project.
   *
   * Reads on the LLM Obs API are eventually consistent, so a pull issued right
   * after a push can momentarily see the dataset missing or only some of its
   * records. This polls with exponential backoff (up to `maxWaitMs`, default 30s)
   * until the dataset is visible and — when `expectedRecordCount` is given — at
   * least that many records are readable, confirming the push fully landed.
   */
  async pullDataset(name: string, options: PullDatasetOptions = {}): Promise<Dataset> {
    const projectId = await this.ensureProjectId();
    const { expectedRecordCount, maxWaitMs = 30_000 } = options;

    let datasetId: string | null = null;
    let description = "";
    let records: DatasetRecord[] = [];
    let recordIds: string[] = [];
    let lastError = "";

    const succeeded = await retryWithBackoff(async () => {
      try {
        // 1. Find the dataset by name.
        if (datasetId === null) {
          const listed = await this._api.listLLMObsDatasets({ projectId, filterName: name });
          for (const item of listed?.data ?? []) {
            if (item?.attributes?.name === name) {
              datasetId = String(item?.id ?? "");
              description = String(item?.attributes?.description ?? "");
              break;
            }
          }
          if (datasetId === null) return false; // not visible yet — keep waiting
        }

        // 2. Fetch its records via the generated client.
        //
        // Spec drift: the API returns record content nested under
        // data[].attributes (JSON:API), but the generated
        // LLMObsDatasetRecordDataResponse model is flat (input/expected_output/
        // metadata at the top level). The deserializer therefore can't map the
        // nested object onto the typed fields (they come back undefined) and
        // instead stashes the whole `attributes` object under
        // `additionalProperties`. Read from there, falling back to the typed
        // fields in case the API/model are realigned later.
        const recs: DatasetRecord[] = [];
        const ids: string[] = [];
        const resp = await this._api.listLLMObsDatasetRecords({ projectId, datasetId });
        for (const item of resp?.data ?? []) {
          const nested = (item as any)?.additionalProperties?.attributes;
          const src = nested ?? item;
          recs.push(
            new DatasetRecord(
              unwrapAnyValue(src?.input) ?? null,
              unwrapAnyValue(src?.expected_output ?? src?.expectedOutput) ?? null,
              (unwrapAnyValue(src?.metadata) as Record<string, unknown>) ?? {},
            ),
          );
          ids.push(String(item?.id ?? ""));
        }
        records = recs;
        recordIds = ids;

        // 3. If a count was expected, keep waiting until everything has landed.
        if (expectedRecordCount != null && recs.length < expectedRecordCount) {
          return false;
        }
        return true;
      } catch (err) {
        lastError = (err as Error).message;
        return false; // transient read error — retry within the budget
      }
    }, { maxTotalMs: maxWaitMs });

    if (datasetId === null) {
      if (lastError) {
        throw new Error(`Failed to list datasets in project '${this._projectName}': ${lastError}`);
      }
      throw new Error(`Dataset '${name}' not found in project '${this._projectName}' (after ${maxWaitMs}ms)`);
    }
    if (!succeeded && expectedRecordCount != null) {
      throw new Error(
        `Dataset '${name}' has ${records.length} record(s) after ${maxWaitMs}ms, expected ${expectedRecordCount} — backend may not have finished ingesting the push`,
      );
    }

    return Dataset.fromExisting(this, name, description, datasetId, projectId, records, recordIds);
  }
}

/**
 * Record `input` / `expectedOutput` are typed `AnyValue` in the generated model;
 * the deserializer wraps some scalar values in an `UnparsedObject` of the form
 * `{ _data: <raw> }`. Unwrap it back to the raw JSON value. (Values read out of
 * `additionalProperties` are already raw and pass through unchanged.)
 */
function unwrapAnyValue(value: unknown): unknown {
  if (value && typeof value === "object" && "_data" in (value as Record<string, unknown>)) {
    return (value as { _data: unknown })._data;
  }
  return value;
}

/** Fluent builder mirroring the Java `ExperimentsClient.Builder`. */
export class ExperimentsClientBuilder {
  private _apiKey = "";
  private _appKey = "";
  private _site = "datadoghq.com";
  private _projectName = "";

  apiKey(apiKey: string): this {
    this._apiKey = apiKey;
    return this;
  }

  applicationKey(appKey: string): this {
    this._appKey = appKey;
    return this;
  }

  site(site: string): this {
    this._site = site;
    return this;
  }

  projectName(projectName: string): this {
    this._projectName = projectName;
    return this;
  }

  build(): ExperimentsClient {
    return new ExperimentsClient({
      apiKey: this._apiKey,
      applicationKey: this._appKey,
      site: this._site,
      projectName: this._projectName,
    });
  }
}
